crates/ruff_linter/src/rules/flake8_type_checking/rules/type_alias_quotes.rs

```diff
@@ -0,0 +1,288 @@
+///
+/// [PEP 604]: https://peps.python.org/pep-0604/
+/// [PEP 613]: https://peps.python.org/pep-0613/
+/// [PEP 695]: https://peps.python.org/pep-0695/#generic-type-alias
+#[violation]
+pub struct QuotedTypeAlias;
+
+impl AlwaysFixableViolation for QuotedTypeAlias {
+    #[derive_message_formats]
+    fn message(&self) -> String {
+        "Remove quotes from type alias".to_string()
+    }
+
+    fn fix_title(&self) -> String {
+        "Remove quotes".to_string()
+    }
+}
+
+/// TCH007
+pub(crate) fn unquoted_type_alias(checker: &Checker, binding: &Binding) -> Option<Vec<Diagnostic>> {
+    if binding.context.is_typing() {
+        return None;
+    }
+
+    if !binding.is_explicit_type_alias() {
+        return None;
+    }
+
+    let Some(Stmt::AnnAssign(ast::StmtAnnAssign {
+        value: Some(expr), ..
+    })) = binding.statement(checker.semantic())
+    else {
+        return None;
+    };
+
+    let mut names = Vec::new();
+    collect_typing_references(checker, expr, &mut names);
+    if names.is_empty() {
+        return None;
+    }
+
+    // We generate a diagnostic for every name that needs to be quoted
+    // but we currently emit a single shared fix that quotes the entire
+    // expression.
+    //
+    // Eventually we may try to be more clever and come up with the
+    // minimal set of subexpressions that need to be quoted.
+    let parent = expr.range().start();
+    let edit = quote_type_expression(expr, checker.semantic(), checker.stylist()).ok();
+    let mut diagnostics = Vec::with_capacity(names.len());
+    for name in names {
+        let mut diagnostic = Diagnostic::new(UnquotedTypeAlias, name.range());
+        diagnostic.set_parent(parent);
+        if let Some(edit) = edit.as_ref() {
+            diagnostic.set_fix(Fix::unsafe_edit(edit.clone()));
+        }
+        diagnostics.push(diagnostic);
+    }
+    Some(diagnostics)
+}
+
+/// Traverses the type expression and collects `[Expr::Name]` nodes that are
+/// not available at runtime and thus need to be quoted, unless they would
+/// become available through `[Rule::RuntimeImportInTypeCheckingBlock]`.
+fn collect_typing_references<'a>(
+    checker: &Checker,
+    expr: &'a Expr,
+    names: &mut Vec<&'a ast::ExprName>,
+) {
+    match expr {
+        Expr::BinOp(ast::ExprBinOp { left, right, .. }) => {
+            collect_typing_references(checker, left, names);
+            collect_typing_references(checker, right, names);
+        }
+        Expr::Starred(ast::ExprStarred {
+            value,
+            ctx: ExprContext::Load,
+            ..
+        })
+        | Expr::Attribute(ast::ExprAttribute { value, .. }) => {
+            collect_typing_references(checker, value, names);
+        }
+        Expr::Subscript(ast::ExprSubscript { value, slice, .. }) => {
+            collect_typing_references(checker, value, names);
+            if let Expr::Name(ast::ExprName { id, .. }) = value.borrow() {
+                if id.as_str() != "Literal" {
```

Comment:
I think we have to make sure here that this is indeed the `typing.Literal` and not just any other literal

Reply:
Good catch, this should indeed use `match_typing_expr`.

Reply:
It also doesn't handle `typing.Annotated` yet. So I will need to fix that as well.